Reconfigure Library

trisongz opened the pull request:
- refactor client into a class
- support `async` operations
- utilizes `httpx` over `requests`
- supports `pydantic` model output as well as `json` outputs
- persist user configurations through json file created at `~/.banana/config.json`
- allow for multiple model inference easily
- simplify logic for inference, as well as passing custom params to `httpx` client
- add validated usage in readme.
- add type-checking across all functions

Commit:
Reconfigure Library

- refactor client into a class
- support `async` operations
- utilizes `httpx` over `requests`
- supports `pydantic` model output as well as `json` outputs
- persist user configurations through json file created at `~/.banana/config.json`
- allow for multiple model inference easily
- simplify logic for inference, as well as passing custom params to `httpx` client
- add validated usage in readme.

erik-dunteman commented:
beautiful! Will take a look real quick

erik-dunteman commented on README.md:
```diff
@@ -10,17 +10,56 @@ Get your API Key
+# Optionally specify the model
+# out = client.run(model_inputs, model_name = 'model1')
+# can also call the client directly:
+# out = client(model_inputs)
```
very clean design

erik-dunteman commented:
Thanks very much for this contribution! 

Got lots of high-level thoughts.

Two main design themes that we'd like to keep with the official SDKs:
1. simplicity over configurability
2. maintaining backward compatibility whenever possible

This is a large PR with breaking changes, so we'll need to fine-tooth comb through it in order to get it in. 

Approved:
- type checking
- httpx
- a BananaClient to pre-set api key and model key
- that slick direct client method `out = client(model_inputs)`

Conditionally Approved:
- async
      - big thanks for doing this, we've been wanting something like this! In my experience with async python is that it almost never plays nicely with sync python. I'd need to see the tests you ran to prove it works both in async and sync contexts, for backward compat.
      
Not Approved:
- Fully replacing .run(), .start(), and .check() functions with the client object. We want to maintain backward compat, so we'd need the BananaClient to be a 100% optional way to call the underlying .run() function. Run, Start, and Check must remain public exported functions for use by existing users.
- Pydactic
     - Our backend is Go based and strongly typed so return types will never be unexpected. Best to keep as a native dict for simplicity.
- Multi-model dictionary
     - also for simplicity; users can build that logic themselves if needing to switch rapidly between models
     
Huge thanks for the contribute. Happy to discuss more here. Adding @kylejmorris and @sahil280114 if they have additional thoughts.

trisongz commented:
Definitely understood, I designed it from an `end-user` perspective rather than an `sdk` first approach. 

`Fully replacing .run(), .start(), and .check() functions with the client object. We want to maintain backward compat, so we'd need the BananaClient to be a 100% optional way to call the underlying .run() function. Run, Start, and Check must remain public exported functions for use by existing users`

I can update the PR tmrw to maintain backwards compat. 

`Pydactic
Our backend is Go based and strongly typed so return types will never be unexpected. Best to keep as a native dict for simplicity`

As mentioned above, I designed this with end-user in mind, and the reason I chose this is because often I work with chained api-calls, and having access to the response model as a python object improves quality of life.

As you mention, end-users can implement this themselves so no reason to keep this if it's not needed.

trisongz commented:
This is the test script I used to validate

```python
import os
import asyncio

os.environ['BANANA_APIKEY'] = '...'

from banana_dev.client import BananaClient

client = BananaClient(
    models = {
        't5-xl-optimized': '....'
    }
)


model_inputs = {'prompt': 'Hello I am a [MASK] model.'}


def test_sync(inputs = model_inputs):
    out = client(inputs)
    print(out)
    return out

async def test_async(inputs = model_inputs):
    out = await client.async_run(inputs)
    print(out)
    return out

async def test_all(inputs = model_inputs):
    test_sync(inputs)
    await test_async(inputs)


if __name__ == '__main__':
    asyncio.run(test_all())

```

Commit:
add backwards compatability layer

Commit:
minor typos

trisongz commented:
Latest commit adds compatibility layer

```python
import os
import asyncio

os.environ['BANANA_APIKEY'] = '...'
# Assuming 
import banana_dev as banana

# method for initialization that can be called once during startup

banana.init_api(models = {'t5-xl-optimized': '...'})

model_inputs = {'prompt': 'Hello I am a [MASK] model.'}


def test_sync(inputs = model_inputs):
    out = banana.run(inputs)
    print(out)
    return out

async def test_async(inputs = model_inputs):
    out = await banana.async_run(inputs)
    print(out)
    return out

async def test_all(inputs = model_inputs):
    test_sync(inputs)
    await test_async(inputs)


if __name__ == '__main__':
    asyncio.run(test_all())


```

Commit:
conformance

- remove `pydantic.BaseModel` dependency
- add `BananaLogs` for simpler log mtg.

erik-dunteman commented:
Thanks for the test scripts and edits! Will take a look and run some tests on my end to repro

erik-dunteman commented:
I've let this PR hang too long and apologize.

We like the Client-object design and plan to implement it across all of our SDKs. We've been tied up with more backend work recently, so rolling out client-object designs to our SDKs has been deprioritized. 

I intend to get this PR merged in the future, but can't promise the standard sub-one-week turnaround other projects have. 

Cool if we keep this open and loop back to it in the future?

jdagdelen commented:
Hey, just saw this thread. I am really interested in having async work. Would it be possible to break this up into smaller chunks (e.g. async separate from other stuff) and merge them individually?
